brainiak/matnorm_models/base.py

```diff
@@ -0,0 +1,229 @@
+
+    def matnorm_logp(self, x, row_cov, col_cov):
+        """Log likelihood for centered matrix-variate normal density.
+        Assumes that row_cov and col_cov follow the API defined in NoiseCovBase.
+        """
+
+        rowsize = tf.cast(tf.shape(x)[0], 'float64')
+        colsize = tf.cast(tf.shape(x)[1], 'float64')
+
+        # precompute sigma_col^{-1} * x'
+        solve_col = col_cov.Sigma_inv_x(tf.transpose(x))
+        logdet_col = col_cov.logdet
+
+        # precompute sigma_row^{-1} * x
+        solve_row = row_cov.Sigma_inv_x(x)
+        logdet_row = row_cov.logdet
+
+        return self._mnorm_logp_internal(colsize, rowsize, logdet_row,
+                                         logdet_col, solve_row, solve_col)
+
+    def matnorm_logp_marginal_row(self, x, row_cov, col_cov, marg, marg_cov):
+        """
+        Log likelihood for centered matrix-variate normal density.
+        Assumes that row_cov, col_cov, and marg_cov follow the API defined
+        in NoiseCovBase.
+
+        When you marginalize in mnorm, you end up with a covariance S + APA',
+        where P is the covariance of A in the relevant dimension.
+
+        This method exploits the matrix inversion and determinant lemmas to
+        construct S + APA' given the covariance API in in NoiseCovBase.
+        """
+        rowsize = tf.cast(tf.shape(x)[0], 'float64')
+        colsize = tf.cast(tf.shape(x)[1], 'float64')
+
+        solve_col = col_cov.Sigma_inv_x(tf.transpose(x))
+        logdet_col = col_cov.logdet
+
+        solve_row, logdet_row = self.solve_det_marginal(x, row_cov, marg,
+                                                        marg_cov)
+
+        return self._mnorm_logp_internal(colsize, rowsize, logdet_row,
+                                         logdet_col, solve_row, solve_col)
+
+    def matnorm_logp_marginal_col(self, x, row_cov, col_cov, marg, marg_cov):
+        """
+        Log likelihood for centered matrix-variate normal density. Assumes that
+        row_cov, col_cov, and marg_cov follow the API defined in NoiseCovBase.
+
+        When you marginalize in mnorm, you end up with a covariance S + APA',
+        where P is the covariance of A in the relevant dimension.
+
+        This method exploits the matrix inversion and determinant lemmas to
+        construct S + APA' given the covariance API in in NoiseCovBase.
+        """
+        rowsize = tf.cast(tf.shape(x)[0], 'float64')
+        colsize = tf.cast(tf.shape(x)[1], 'float64')
+
+        solve_row = row_cov.Sigma_inv_x(x)
+        logdet_row = row_cov.logdet
+
+        solve_col, logdet_col = self.solve_det_marginal(tf.transpose(x),
+                                                        col_cov, marg, marg_cov)
+
+        return self._mnorm_logp_internal(colsize, rowsize, logdet_row,
+                                         logdet_col, solve_row, solve_col)
+
+    def matnorm_logp_conditional_row(self, x, row_cov, col_cov, cond, cond_cov):
```

Comment:
If they stay as methods in MatnormBase, agreed. But see above about possibly moving them out, in which case I think they're fine returning TF objects (possibly even following tf.contrib.distributions API).